### PR TITLE
Add check for reference-compared literals to JS files

### PR DESCRIPTION
**Pull request by Jack-Works**
I don't know where I can add a test for JS files.

According to discussion in https://github.com/microsoft/TypeScript/pull/45978, I open this PR to also check JavaScript files.

### CODE REVIEW

**Comment by sandersn** [comment]
Plain JS tests are in `$ts/tests/cases/conformance/salsa/plainJS*`. You can get the settings you need to test them from the existing tests.

A couple of gotchas I thought of:
- `"[object Object]" == {}`
- `{ toString() { return 'hi' }, valueOf() { return 'no' } } == { toString() { return 'no' }, valueOf() { return 'hi' } }`

At least, I think there are shenanigans possible with `valueOf`; I couldn't get a working example.

**Comment by Jack-Works** [comment]
I think there's no need to support the first case. The second case is a bit more interesting, maybe I can add detection to check if the literal contains `toString`, `valueOf`, or any computed property name only in JS.

**Comment by Jack-Works** [comment]
Add test files.

**Comment by Jack-Works** [comment]
So we only emit it for `===` not for `==`?

**Comment by sandersn** [comment]
Yeah, I think so.

**Comment by Jack-Works** [comment]
done

**Comment by Jack-Works** [comment]
rebased

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -34767,8 +34767,13 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             case SyntaxKind.EqualsEqualsEqualsToken:
             case SyntaxKind.ExclamationEqualsEqualsToken:
                 if (isLiteralExpressionOfObject(left) || isLiteralExpressionOfObject(right)) {
-                    const eqType = operator === SyntaxKind.EqualsEqualsToken || operator === SyntaxKind.EqualsEqualsEqualsToken;
-                    error(errorNode, Diagnostics.This_condition_will_always_return_0_since_JavaScript_compares_objects_by_reference_not_value, eqType ? "false" : "true");
+                    if (isInJSFile(left) && (operator === SyntaxKind.EqualsEqualsToken || operator === SyntaxKind.ExclamationEqualsToken)) {
+                        // we skip the emit for {} == {} in JS files
+                    }
+                    else {
```
Comment (sandersn) [inline src/compiler/checker.ts:34773]: invert this check, and no comment is needed
```suggestion
                    if (!isInJSFile(left) || operator !== SyntaxKind.EqualsEqualsToken && operator !== SyntaxKind.ExclamationEqualsToken) {
```

**Review by sandersn — CHANGES_REQUESTED** [review]
Almost ready, just one more change.

It's been a long time, so if you want me to finish this, let me know.

File: src/compiler/checker.ts (hunk 1 above)
Comment (sandersn) [reply src/compiler/checker.ts:34773]: actually, merge this check with the previous one

**Review by sandersn — APPROVED** [review]

**Pull request merged**